hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -2245,24 +2248,32 @@ class Emit[C](
 
             Code.foreach(requestedShape) { requestedShapeElement => Code(
               tempShapeElement := requestedShapeElement,
-              (tempShapeElement <= 0L).mux(
+              (tempShapeElement < 0L).mux(
                 (tempShapeElement ceq -1L).mux(
                   hasNegativeOne.mux(
                     Code._fatal[Unit]("Can't infer shape, more than one -1"),
                     hasNegativeOne := true
                   ),
-                  Code._fatal[Unit]("Can't reshape, new shape must contain only positive numbers or -1")),
+                  Code._fatal[Unit]("Can't reshape, new shape must contain only nonnegative numbers or -1")),
                 runningProduct := runningProduct * tempShapeElement
               )
             ) },
-            hasNegativeOne.mux(
-              (numElements % runningProduct) > 0L,
-              numElements cne runningProduct
-            ).orEmpty(Code._fatal[Unit]("Can't reshape since requested shape is incompatible with number of elements")),
-            quotient := numElements / runningProduct,
-            Code(newShapeVars.zip(requestedShape).map { case (variable, shapeElement) =>
-              variable := (shapeElement ceq -1L).mux(quotient, shapeElement)
-            })
+            (runningProduct ceq 0L).mux(
```

Comment:
I think the logic here could be cleaned up a bit. How about
```
Code(
  hasNegativeOne.mux(
    (runningProduct ceq 0) || (numElements % runningProduct) > 0L,
    numElements cne runningProduct
  ).orEmpty(Code._fatal[Unit]("Can't reshape since requested shape is incompatible with number of elements")),
  quotient := (runningProduct ceq 0).mux(0, numElements / runningProduct),
  Code(newShapeVars.zip(requestedShape).map { case (variable, shapeElement) =>
    variable := (shapeElement ceq -1L).mux(quotient, shapeElement)
  })
```